_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,499 @@
+
+```ruby
+class User < ActiveRecord::Base
+  # Schema
+  # Name: string
+  # UUID: UUID (primary key)
+  # Email
+  # Devise password stuff
+  devise :database_authenticatable
+
+  def encrypted
+    cipher         =  OpenSSL::Cipher.new('aes-256-cbc')
+    cipher.encrypt
+    cipher.key     =  ENV['SOME_SECRET_KEY']
+    cipher.iv      =  ENV['SOME_SECRET_IV']
+    cipher.padding =  1
+    encrypted      =  cipher.update(self.uuid) + cipher.final # use only this user's UUID for encryption
+    Base64.urlsafe_encode64(encrypted).encode('utf-8')
+  end
+end
+```
+
+Look at this bonkers method! It's pretty hard to figure out what's going on here. And to be fair it's outside 
+the norm of ordinary, MVC style web development. So lets go through it line by line.
+
+
+#### Line 1
+```ruby
+cipher =  OpenSSL::Cipher.new('aes-256-cbc')
+```
+
+The argument 'aes-256-cbc' is a hyphenated description of components of the cipher. 
+
+1. "aes" is the name of the encryption algorithm. In this case, its an acronym for [Advanced Encryption Standard](http://aesencryption.net/).
+
+2. 256 is the number of bits of the key used in the algorithm. You'll usually see 128 or 256. Practically, this means that the key is "large" enough that it cannot be feasibly broken via a brute force attack (it would take foooooorrreeeevvvveeeerrr) . 
+
+3. cbc stands for "cipher block chaining" and is the mode by which the aes algorithm encrypts the data. Its most commonly compared to the ecb [electronic code book](http://searchsecurity.techtarget.com/definition/Electronic-Code-Book). 
+Its generally preferred to use cbc because ebc may expose pieces of the actual key in the encrypted string when encrypting smaller pieces of data. 
+
+Anyway, like electricity, you needn't full understand these modes to use them.
+But full documentation is always [available](http://docs.ruby-lang.org/en/trunk/OpenSSL/Cipher.html). 
+
+---
+
+#### Line 2
+```ruby
+cipher.encrypt # cipher.decrypt
+```
+
+Since you instantiate a "decipher" the same as a cipher, this line lets the instance know that we intend to use it for encryption rather than decryption.
+
+---
+
+#### Line 3
+```ruby
+cipher.key =  ENV['SOME_SECRET_KEY']
+```
+
+OK, here's where it starts to get interesting. 
+This is the first part of the password that we will shared between applications. 
+Its important to create a truly random (or as close to that as possible) key, so consult the [documentation](http://docs.ruby-lang.org/en/trunk/OpenSSL/Cipher.html#class-OpenSSL::Cipher-label-Choosing+a+key)
+to get something as secure as possible.
+
+---
+
+#### Line 4
+```ruby
+cipher.iv =  ENV['SOME_SECRET_IV']
+```
+
+IV, well this stands for initialization vector. An initialization vector creates an extra layer of security around a key, and is generally used only once per instance of a cipher. 
+We're breaking the rules a little here, because our IV will be used only to prevent the accidental exposure of the key in the encrypted payload. 
```

Comment:
DOODS, can this stuff be in the article? because I didn't know this, and that's super interesting (the discussion about generating a separate IV for each user)